text/3113-bitfields.md

```diff
@@ -0,0 +1,541 @@
+- Feature Name: bitfield
+- Start Date: 2021-04-22
+
+- RFC PR: [rust-lang/rfcs#3113](https://github.com/rust-lang/rfcs/pull/3113)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000) 
+
+# Summary
+[summary]: #summary
+
+This RFC adds support for bit-fields in `repr(C)` structs by
+
+- Introducing a new attribute `bits(N)` that can be applied to integer
+  fields.
+- Allowing such annotated fields to be unnamed.
+
+# Motivation
+[motivation]: #motivation
+
+The Linux kernel user-space API contains over 400 bit-fields. Writing the
+corresponding types in Rust poses significant problems because the layout of
+structs that contain bit-fields varies between architectures and is hard to
+calculate manually. Consider the following examples which were compiled with
+Clang for the `X-unknown-linux-gnu` target:
```

Comment:
This brings up a question I have been meaning to ask: aren't C bitfields highly compiler-defined? What parts of bitfield allocation are defined by the implementation of the compiler, and why can we trust the bitfields designed for C interop actually match the implementation for that compiler? I do not believe this is a trivial implementation question we can shrug off for later.

Reply:
Off the top of my head I have seen compilers change the following between implementations (at least where important for interoperability). Some implementations even allow these to be configured either through compiler flags or `#pragma`s.

- default bit order (MSB vs LSB)
- what integer types are allowed (by default only `int` and `unsigned int` are allowed)
- bit packing (how the fields as defined actually get squished into the bytes)
- if fields can straddle storage-unit boundaries
  - e.g. with the following bitfield, will the size of the structure be two bytes or three? Assuming LSB allocation order, will `b` include bits `6..=9` or `8..=11`?
  ```
  struct example_t {
      unsigned char a: 6;
      unsigned char b: 4;
      unsigned char c: 6;
  }
  ```

Based on all of this I think if Rust is going to support bitfields natively, we need to select one set of features and stick to it.

Reply:
Well, that's... somewhat stressful to think about. It sounds like we may have trouble actually finding a rule to adhere to for the Rust compiler that produces a conformant implementation. I found the following in C11, 6.7.2.1:

11. An implementation may allocate any addressable storage unit large enough to hold a bit-field. If enough space remains, a bit-field that immediately follows another bit-field in a structure shall be packed into adjacent bits of the same unit. If insufficient space remains, whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is implementation-defined. The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined. The alignment of the addressable storage unit is unspecified.
12. A bit-field declaration with no declarator, but only a colon and a width, indicates an unnamed bit-field.) As a special case, a bit-field structure member with a width of 0 indicates that no further bit-field is to be packed into the unit in which the previous bit-field, if any, was placed.

The System V AMD64 ABI adds:
- bit-fields are allocated from right to left
- bit-fields may share a storage unit with other struct / union member
- Unnamed bit-fields’ types do not affect the alignment of a structure or union.

And I have taken a selection of things from the Arm64 AAPCS here (it includes a reasonably complete layout calculation algorithm, which I have not quoted in full):
- For each bit-field, the type of its container is:
  - Its declared type if its size is no larger than the size of its declared type.
  - The largest integral type no larger than its size if its size is larger than the size of its declared type (see Over-sized bit-fields).
- The container type contributes to the alignment of the containing aggregate in the same way a plain (not bit-field) member of that type would, without exception for zero-sized or anonymous bit-fields.
- The content of each bit-field is contained by **exactly one** instance of its container type.
- For big-endian data types K(F) is the offset from the most significant bit of the container to the most significant bit of the bit-field.
- For little-endian data types K(F) is the offset from the least significant bit of the container to the least significant bit of the bit-field.
- The AAPCS **does not** allow exported interfaces to contain packed structures or bit-fields.

The last one is kind of funny because it makes me wonder if that means we should compile error if someone tries to yield a struct with a bitfield to `extern "C"` on Arm?

Reply:
Apparently the issue, as explained to me on Zulip, is actually that "packed structures" or "packed bit-fields" may not be used through "exported interfaces". Here an "exported interface" is anything that gets exposed to _other_ programs via symbols, etc. in the usual manner and thus may have a call site that is not directly controlled by the compiler in question.

For those procedures exclusively governed by a single compiler, of course. it can ignore the entire AAPCS anyways, since it is the boundaries between programs (or libraries, etc.) that are what the AAPCS is designed to govern.